File: tests/test_move_arg_bp.py
import pytest
from test_move_arg_bp import Item, access, access_ptr, consume


def test():
    item = Item(42)
    other = item
    print(item)
    access(item)
    consume(item)
    print("back in python")

    try:
        access_ptr(item)
        access(item)
    except Exception as e:
        print(e)

    del item

    try:
        print(other)
    except Exception as e:
        print(e)

if __name__ == "__main__":
    test()

File: tests/test_move_arg_bp.cpp
#include "test_move_arg.h"
#include <boost/python.hpp>
#include <sstream>

namespace py = boost::python;

std::string item_repr(const Item& item) {
	std::stringstream ss;
	ss << "py " << item;
	return ss.str();
}

void item_access(const Item& item) {
	std::cout << "access " << item << "\n";
}

void item_access_ptr(const std::auto_ptr<Item>& item) {
	std::cout << "access ptr " << item.get() << " ";
	if (item.get()) std::cout << *item;
	std::cout << "\n";
}

void item_consume(std::auto_ptr<Item>& item) {
	std::cout << "consume " << *item << "\n  ";
	Item sink(std::move(*item.release()));
	std::cout << "  old: " << item.get() << "\n  new: " << sink << "\n";
}

BOOST_PYTHON_MODULE(test_move_arg_bp) {
	py::class_<Item, std::auto_ptr<Item>, boost::noncopyable>("Item", py::init<int>())
		.def("__repr__", &item_repr);

	py::def("access", item_access);
	py::def("access_ptr", item_access_ptr);

	py::def("consume", item_consume);
}
